### ISSUE TITLE
Check, at compile time, that ReturnRef action is not operating on temporaries 

### ISSUE THREAD

**Issue by PiotrNycz** [google/googletest#2471]
In the below example `::testing::ReturnRef` store reference to temporary:

```
struct SomeClass { virtual void a() const {} };
struct Interface {  virtual const SomeClass& foo() = 0; };
void codeUnderTest(Interface& interface) {
    interface.foo().a();
}

struct InterfaceMock : Interface { MOCK_METHOD0(foo, const SomeClass&()); };

struct TestCode : ::testing::Test {
      InterfaceMock mock;
      static SomeClass foo1() {   return {}; }
};

TEST_F(TestCode, testFoo)
{
      using namespace ::testing;
      EXPECT_CALL(mock, foo()).WillRepeatedly(ReturnRef(foo1())); // DANGLING REFERENCE

       codeUnderTest(mock);
}

```

The issue can be easily fixed by adding, next to ReturnRef function, the deleted version of it for r-value references:

```

template <typename T>
auto ReturnRef(T& p_arg) {....}

// added code
template <typename T> 
void ReturnRef(T&&) = delete;

```

Fortunately the issue is quite rare because `T&&` cannot be cast to `T&` if `T` is not const type. So in the above example - the return value of problematic function is `const SomeClass &`, not `SomeClass &`.
OF course - the issue is not existent also for pre-C++11 code.

**Comment by kuzkry** [comment]
Hello!
Please, before submitting something please make sure your example code compiles.
The current googletest implementation looks good enough for me due to [temporary object lifetime](https://en.cppreference.com/w/cpp/language/lifetime#Temporary_object_lifetime).
As you noticed, T&& cannot be cast to T& if T is not const type so no problem here either.
In your case, try removing `const` specifier from `static const A foo1() {   return A{}; }` - and this, btw. is something I would consider a user-side mistake.

**Comment by PiotrNycz** [comment]
@kuzkry  valid point about not compilable code - I've corrected it.

For you second point. If the `const` was removed from `static const SomeClass foo1() { return {}; }` - we actually get this message (googletest version 1.8, gcc7.x, C++17):

> .../SomeTest.cpp: In member function 'virtual void TestCode_testFoo_Test::TestBody()':
> ../SomeTest.cpp:17:61: error: cannot bind non-const lvalue reference of type 'SomeClass&' to an rvalue of type 'SomeClass'
>        EXPECT_CALL(mock, foo()).WillRepeatedly(ReturnRef(foo1())); // DANGLING REFERENCE
>                                                          ~~~~^~
> In file included from D:/workspace/googletest-master/googlemock/include/gmock/gmock.h:58:0,
>                  from .../SomeTest.cpp:2:
> D:/workspace/googletest-master/googlemock/include/gmock/gmock-actions.h:1077:37: note:   initializing argument 1 of 'testing::internal::ReturnRefAction<R> testing::ReturnRef(R&) [with R = SomeClass]'
>  inline internal::ReturnRefAction<R> ReturnRef(R& x) {  
>                                      ^~~~~~~~~
> 

So - in many cases - user will add const next to return value - and we got dangling references w/o any compilation errors. So, I think, my proposal is valid - because it shows to the user the real issue here.

**Comment by PiotrNycz** [comment]
@kuzkry 

There is another dangerous "workaround" - using explicit specialization to get rid of current errors:  `ReturnRef<const SomeClass>(foo1())`. This code compiles fine - and have dangling reference.

**Comment by kuzkry** [comment]
I mean I don't think we have any dangling references at all because of _temporary object lifetime_:
> Temporary objects are created (...)  in the following situations:
binding a reference to a prvalue
...

You may also add `-fsanitize=undefined`, which was clean in my tests. Please correct me if I'm wrong.

**Comment by PiotrNycz** [comment]
@kuzkry  "-fsanitize=address" catches this error (I have to switch to linux to use it) on this very example code. 
But I have in my huge (few MLOC) project cases when address sanitizer was not able to catch that. 

And the proposed change, that I made to gmock locally, allowed to catch it. 

And - I am not saying in this issue there is anything wrong with gtest/gmock code. I am just proposing to  add, maybe as an optional feature - switched ON/OFF by preprocessor macro, this few lines code, to allows user get such kind of errors more easily, and more predictable, than using address sanitizer or other such tools. I mean - during compilation, not during runtime.

**Comment by kuzkry** [comment]
Oh, indeed, you're right! Sorry for that :)
I've applied your patch like this:
```c++
template <typename R>
inline internal::ReturnRefAction<R> ReturnRef(R&& x) = delete;
```
and I can tell your fix seems to be enough to solve the additional problem you have mentioned, i.e.:
> There is another dangerous "workaround" - using explicit specialization to get rid of current errors: ReturnRef<const SomeClass>(foo1()). This code compiles fine - and have dangling reference.

Because explicit specialization `ReturnRef<const SomeClass>(foo1())` tries to call the deleted version (tested on GCC and Clang) - because the deleted version is a better match. But the patch doesn't work for function pointers though.
```c++
using Ret = decltype(ReturnRef(std::declval<SomeClass const&>()));
using Arg = SomeClass const&;
using Fn = Ret(Arg);
using FnPtr = Fn*;
FnPtr fn_ptr = &ReturnRef; // it will bind, not quite what we want!

fn_ptr(foo1()); // dangling reference
```
Perhaps we have no language support to fully resolve this kind of issues or I have yet to come up with something else.

**Comment by kuzkry** [comment]
Okay, I went to read C++ standard a bit and yes, this const qualifier in the return type changes a lot. About my function pointer issue - well, as you see we may as well say I sort of lied to the compiler, so it's an user-side mistake we probably cannot prevent (maybe someone smarter than me can figure it out). Thanks for the report anyway :)

Nevertheless, would you kindly prepare a pull request for that?

**Comment by PiotrNycz** [comment]
@kuzkry  I can try in next week. Is there any "howto" for pull-requests - what tests shall I run, etc?

**Comment by kuzkry** [comment]
I think I should have explained myself better. I didn't intent to bother you with this, I simply encourage you to contribute (but you can hand it over if you want).
Yes, we've got our [tutorial](https://github.com/google/googletest/blob/master/CONTRIBUTING.md), please read it.
Once you create a pull request, it will also undergo CI verification. Since that's gonna be your first contribution, you will also need to sign the CLA of Google.

**Comment by kuzkry** [comment]
@PiotrNycz, any update? Has something hold you up?

**Comment by PiotrNycz** [comment]
@kuzkry I hope I did this merge-request according to gtest/gmock rules. If not - go do it by yourself - I would not mind. 
PS. I do not have strong excuses for this delay - just have more urgent things to do.

**Comment by kuzkry** [comment]
Yes, it looks good. In this particular case, you won't need anything else.

**Comment by kuzkry** [comment]
Merged by #2527. Please close this issue down.
